File: hackerrank/problem solving/algorithms/2 - implementation/13 - counting valleys/counting_valleys.py
# https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

def main():
    input()
    steps = input()
    print(counting_valleys(steps))


# n: length of steps
# T: O(n)
# S: O(1) extra space
def counting_valleys(steps):
    valleys = 0
    current_altitude = 0

    for step in steps:
        was_below_sea_level = current_altitude < 0
        current_altitude += -1 if step == 'D' else 1
        is_in_sea_level_from_valley = was_below_sea_level and current_altitude == 0
        if is_in_sea_level_from_valley:
            valleys += 1

    return valleys


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/13 - counting valleys/counting_valleys.js
// https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

function main() {
  readLine();
  const steps = readLine();
  console.log(countingValleys(steps));
}

// n: length of steps
// T: O(n)
// S: O(1) extra space
function countingValleys(steps) {
  let valleys = 0;
  let currentAltitude = 0;

  for (const step of steps) {
    const wasBelowSeaLevel = currentAltitude < 0;
    currentAltitude += step === 'D' ? -1 : 1;
    const isInSeaLevelFromValley = wasBelowSeaLevel && currentAltitude === 0;
    if (isInSeaLevelFromValley) valleys++;
  }

  return valleys;
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}
